[SDTEST-901] Auto instrumentation docs for Test Optimization in Ruby (#26648)
DataDog/documentation: content/en/tests/setup/ruby.md

| Test runner | Version |
|---|---|
| Knapsack Pro | >= 7.2.0 |
| parallel_tests | >= 4.0.0 |
| ci-queue | >= 0.53.0 |

## Configuring reporting method

To report test results to Datadog, you need to configure the `datadog-ci` gem:

{{< tabs >}}
{{% tab "CI Provider with Auto-Instrumentation Support" %}}
{{% ci-autoinstrumentation %}}

<div class="alert alert-warning">

>   <strong>Note</strong>: Auto-instrumentation is not supported for JRuby. If you use JRuby, please follow the manual instrumentation steps outlined below.

Reviewer: ```suggestion
  <strong>Note</strong>: Auto-instrumentation is not supported for JRuby. Follow the manual instrumentation steps outlined below instead.
```

> ## Instrumenting your tests

Author: This is still auto instrumentation, just the one without CI provider integration. The real manual instrumentation doc is down below. Do you have any suggestions for a better title that means "Instrumenting your tests when your CI provider is not supported but still using auto-instrumentation with no code changes"? :sweat

> ## Instrumenting your tests

Reviewer: @anmarchenko Oh sorry, misunderstood a bit. I will update my suggestion.

Follow these steps if your CI Provider is not supported for auto-instrumentation (see [Configuring reporting method](#configuring-reporting-method)).

1. Set the following environment variables to configure the tracer:

`DD_CIVISIBILITY_ENABLED=true` (Required)
: Enables the Test Optimization product.

`DD_ENV` (Required)
: Environment where the tests are being run (for example: `local` when running tests on a developer workstation or `ci` when running them on a CI provider).

> `DD_SERVICE` (Required)

Reviewer 2: As we're transitioning to the repo-centric view, I think it makes sense to mark the service name as optional

Author: true

```ruby
if ENV["DD_ENV"] == "ci"
  Datadog.configure do |c|
    #  ... ci configs and instrumentation here ...
    c.tracing.instrument :redis
    c.tracing.instrument :pg
    # ... any other instrumentations supported by datadog gem ...
  end
end
```

Alternatively, you can enable automatic APM instrumentation in `test_helper/spec_helper`:

```ruby
require "datadog/auto_instrument" if ENV["DD_ENV"] == "ci"
```

**Note**: In CI mode, these traces are submitted to Test Optimization, and they do **not** show up in Datadog APM.

For the full list of available instrumentation methods, see the [tracing documentation][6]

## Collecting Git metadata

{{% ci-git-metadata %}}

## Manual instrumentation (optional)

<div class="alert alert-info">
<strong>Attention</strong>: when using manual instrumentation, run your tests like you normally do: 
don't change `RUBYOPT` env variable and don't prepend `bundle exec ddcirb exec` to your test command
</div>

Auto instrumentation adds additional performance overhead on code loading stage. It can be especially noticeable for 
very big repositories with a lot of dependencies. If your project takes 20+ seconds to start, you will likely

> benefit from manually instrumenting your tests.

Reviewer: ```suggestion
to benefit from manually instrumenting your tests.
```
